drivers/platform/stm32/stm32_irq.c

```diff
@@ -0,0 +1,278 @@
+ *    patent holders to use this software.
+ *  - Use of the software either in source or binary form, must be run
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+#include <stdlib.h>
+#include <errno.h>
+#include "util.h"
+#include "irq.h"
+#include "stm32_irq.h"
+#include "gpio.h"
+#include "stm32_gpio.h"
+
+/**
+* @struct stm32_irq_callback_desc
+* @brief Structure holding the callback functions for STM32 irqs
+* @note  The callback functions are mapped with 'irq_id' structure
+*/
+typedef struct stm32_irq_callback_desc {
+	uint16_t GPIO_Pin;
+	struct callback_desc desc;
+} stm32_irq_callback_desc[NB_INTERRUPTS];
+
+stm32_irq_callback_desc stm32_irq_callback;
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief STM32 GPIO external interrupt callback function
+ * @param GPIO_Pin - GPIO pin
+ * @return none
+ */
+void HAL_GPIO_EXTI_Callback(uint16_t GPIO_Pin)
+{
+	enum irq_id id;
+
+	/* Invoke the callback function attached to corresponding GPIO pin */
+	for (id = EXTERNAL_INT_ID1; id <= EXTERNAL_INT_ID5; id++) {
+		if (stm32_irq_callback[id].GPIO_Pin == GPIO_Pin) {
+			stm32_irq_callback[id].desc.callback(stm32_irq_callback[id].desc.ctx,
+							     id, NULL);
+			return;
+		}
+	}
+}
+
+/**
+ * @brief Initialize the STM32 interrupt
+ * @param desc - IRQ descriptor
+ * @param param - IRQ initialization parameters
+ * @return SUCCESS in case of success, FAILURE otherwise
+ */
+int32_t stm32_irq_ctrl_init(struct irq_ctrl_desc **desc,
+			    const struct irq_init_param *param)
+{
+	struct irq_ctrl_desc *descriptor;
+	struct stm32_irq_desc *extra;
+	int32_t ret;
+
+	if (!desc || !param)
+		return -EINVAL;
+
+	descriptor = (struct irq_ctrl_desc *)malloc(sizeof(*descriptor));
+	if (!descriptor)
+		return -ENOMEM;
+
+	extra = (struct stm32_irq_desc*)malloc(sizeof(*extra));
+	if (!extra) {
+		ret = -ENOMEM;
+		goto error_desc;
+	}
+
+	switch (param->irq_ctrl_id) {
```

Comment:
Hi @buha 
Sorry for coming very late on this PR.

Actually I was following the ADuCM3029 irq drivers for STM32 irq driver development. 
1.If you look at the 'aducm3029_irq_ctrl_init()' function from 3029, the 'irq_ctrl_id' is referenced from the 'irq_init_param' structure declared in 'irq.h' include file. That means each platform specific 'irq_extra.h' file will contain interrupt number constants (enum irq_id) which will be assigned to 'irq_ctrl_id' param from 'irq_init_param' structure. So code from lines 107-114 make sense to me as 'irq_id' is referenced from high level (generic) irq init param structure and not the stm32 specific irq init extra structure.
2.Also the 'irq_ctrl_id' is copied into 'desc' but is never used again in 3029. The reason being, each irq function from .c file has explicit 'irq_id' argument, so this irq_id is redundant at later stages. Same thing has been done for STM32 irq drivers I believe.

Please let me know if this make sense and if I understood your original comment.

Reply:
Hi @mphalke ,
Sorry, I missed your comment here.
So we've been going through some thinking and rethinking of how irq platform driver should look like and one colleague came up with this: https://github.com/analogdevicesinc/no-OS/pull/1113
This is quite a long PR discussion because the topic is not trivial.
I think this PR should now be the standard for irq platform drivers and here's an example of the usage:
https://github.com/analogdevicesinc/no-OS/blob/58b25b627e56278d5731aaac4e6506ceb3218c70/projects/ad4110/src/app/main.c#L98-L112
https://github.com/analogdevicesinc/no-OS/blob/master/drivers/afe/ad4110/ad4110.c#L871-L886

Reply:
Hi @buha 
Thanks for the comment. Sorry it's not clear to me what change is expected in STM32 irq platform drivers. I looked it over and over again and find that all the functions defined in this file are matching to ones defined in drivers/api/irq.c file and platform ops structure except addition of new function: irq_trigger_level_set(), which is not currently implemented in stm32 irq drivers. Is it expected that this function should be implemented into STM32 irq module because by looking at AD4110 drivers, the irq ID is now passed as GPIO number for which trigger needs to be set. So seems like IRQ ID should not be a hardcoded into platform specific extra file and rather it is attached to GPIO instance?
Could you please throw some light on this?

Reply:
Hi @mphalke,

To throw some light, `#define GPIO_IRQ_ID 	52U` refers to GPIO's interrupt ID to be used by the generic interrupt controller (used by Xilinx and not by stm32).  GPIO's interrupt ID is not equal to number of the pin. 

![image](https://user-images.githubusercontent.com/89378331/145411640-42c41c4f-54e0-471c-86f5-f194b228cd24.png)